pallets/oracle-feed/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+//! Oracle pallet to feed values.
+//! Feeding is permissionless given an initial fee for each key.
+//! This pallet do not aggregate/validate any, it just store them by account as
+//! they come. It's expected other pallet read this storage and
+//! aggregate/validate the values.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+pub mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use cfg_traits::{fees::PayFee, ValueProvider};
+	use frame_support::{pallet_prelude::*, traits::Time};
+	use frame_system::pallet_prelude::*;
+
+	use crate::weights::WeightInfo;
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	type MomentOf<T> = <<T as Config>::Time as Time>::Moment;
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Identify an oracle value
+		type OracleKey: Parameter + Member + Copy + MaxEncodedLen;
+
+		/// Represent an oracle value
+		type OracleValue: Parameter + Member + Copy + MaxEncodedLen + Default;
+
+		/// A way to obtain the current time
+		type Time: Time;
+
+		/// Fee for the first time a feeder feeds a value
+		type FirstValuePayFee: PayFee<Self::AccountId>;
+
+		/// The weight information for this pallet extrinsics.
+		type WeightInfo: WeightInfo;
+	}
+
+	/// Store all oracle values indexed by feeder
+	#[pallet::storage]
+	pub(crate) type FedValues<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		Option<T::AccountId>,
+		Blake2_128Concat,
+		T::OracleKey,
+		(T::OracleValue, MomentOf<T>),
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		Fed {
+			account_id: Option<T::AccountId>,
+			key: T::OracleKey,
+			value: T::OracleValue,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Permissionles call to feed an oracle key from a source with value.
+		/// The first time the value is set, an extra fee is required for the
+		/// feeder.
+		#[pallet::weight(T::WeightInfo::feed_with_fee())]
+		#[pallet::call_index(0)]
+		pub fn feed(
+			origin: OriginFor<T>,
+			key: T::OracleKey,
+			value: T::OracleValue,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed_or_root(origin)?;
```

Comment:
This needs to also allow any other origin. E.g. Bridge origin if we receive feeds from other domains or so.

Reply:
Mmm... Ok, I understand what you want now. So, that should be checked by another `T::EnsureFeeder` type or similar the runtime later configures

Reply:
mmh. I think all but `None` are fine. But that should be a runtime parameter probably ^^

Reply:
Finally by an `EnsureOrigin` parameter, which is configured as you say